File: Stack/907_Sum of Subarray Minimums.py
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        stack = []
        res = 0
        MOD = 10 ** 9 + 7

        for i in range(len(arr) + 1):
            while stack and (i == len(arr) or arr[stack[-1]] >= arr[i]):
                mid = stack.pop()
                left = -1 if not stack else stack[-1]
                right = i
                count = (mid - left) * (right - mid)
                res += arr[mid] * count
            stack.append(i)
        
        return res % MOD

File: Java Solutions/907_SumOfSubarrayMinimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        long res = 0;
        Stack<Integer> stack = new Stack<>();
        int MOD = 1000000007;

        for (int i = 0; i <= arr.length; i++) {
            while (!stack.empty() && (i == arr.length || arr[stack.peek()] >= arr[i])) {
                int mid = stack.pop();
                int left = stack.empty() ? -1 : stack.peek();
                int right = i;
                long count = (mid - left) * (right - mid);
                res += arr[mid] * count;
            }
            stack.push(i);
        }
        
        return (int) (res % MOD);
    }
}